### PR TITLE
Motion Matching: Unifying joint velocity calculations

### PR DESCRIPTION
**Pull request by amzn-jillich**
* Unified the way to calculate joint velocities for the feature (based on the motion database) as well as calculating the joint velocities of the current character pose (query pose) which makes the code easier to maintain and actually also fixed an issue resulting in smoother and better animations.
* Removed helper functions to calculate joint velocities from the base feature class.

Signed-off-by: Benjamin Jillich <jillich@amazon.com>

### CODE REVIEW

File: Gems/MotionMatching/Code/Source/FeatureVelocity.cpp [hunk 1]
```diff
@@ -56,10 +69,10 @@ namespace EMotionFX::MotionMatching
         const Transform relativeToWorldTM = pose->GetWorldSpaceTransform(relativeToJointIndex);
 
         const AZ::Vector3 jointPosition = relativeToWorldTM.TransformPoint(jointModelTM.m_position);
-        const float scale = 0.15f;
-        const AZ::Vector3 velocityWorldSpace = relativeToWorldTM.TransformVector(velocity * scale);
+        const AZ::Vector3 velocityWorldSpace = relativeToWorldTM.TransformVector(velocity);
 
-        DebugDrawVelocity(debugDisplay, jointPosition, velocityWorldSpace, color);
+        const float scale = 0.1f;
```
Comment (hultonha) [inline Gems/MotionMatching/Code/Source/FeatureVelocity.cpp:74]: suggestion/question: Is it worth exposing this as an AZ_CVar if we'd ever want to change this to make it easier to see or reduce to hide it?

File: Gems/MotionMatching/Code/Source/PoseDataJointVelocities.cpp [hunk 2]
```diff
@@ -129,78 +129,82 @@ namespace EMotionFX::MotionMatching
...
+                const float scale = 0.1f;
+                DebugDrawVelocity(debugDisplay, jointPosition, velocityWorldSpace * scale, color);
             }
         }
     }
 
-    void PoseDataJointVelocities::CalculateVelocity(MotionInstance* motionInstance, size_t relativeToJointIndex)
+    void PoseDataJointVelocities::CalculateVelocity(const ActorInstance* actorInstance, Motion* motion, float requestedSampleTime, size_t relativeToJointIndex)
     {
+        MotionDataSampleSettings sampleSettings;
```
Comment (hultonha) [inline Gems/MotionMatching/Code/Source/PoseDataJointVelocities.cpp:142]: suggestion: Is it worth zero initializing this to make sure the fields are zeroed? `MotionDataSampleSettings sampleSettings{};`

File: Gems/MotionMatching/Code/Source/PoseDataJointVelocities.cpp [hunk 3]
```diff
@@ -129,78 +129,82 @@ namespace EMotionFX::MotionMatching
...
                 const AZ::Vector3 velocity = CalculateLinearVelocity(prevPosition, currentPosition, frameDelta);
-                m_velocities[jointIndex] += inverseJointWorldTransform.TransformVector(velocity);
+                m_velocities[jointIndex] += velocity;
             }
 
             *prevPose = *currentPose;
         }
 
-        const float numSamplesFloat = aznumeric_cast<float>(numSamples);
+        const float numSamplesFloat = aznumeric_cast<float>(numSamples - 1);
```
Comment (hultonha) [inline Gems/MotionMatching/Code/Source/PoseDataJointVelocities.cpp:201]: Good spot! 😅

**Review by hultonha — APPROVED** [review]
Change LGTM! I realize something like this is difficult to test but it would be really cool to maybe even experiment with a 'golden master' style approach where we run the simulation with fixed inputs for a set period of time and record the output (which we know is correct) and then ensure the current output in the test continues to match that (likely within some tolerance). That way if this code is updated/refactored in future we might be alerted more quickly to an issue that is introduced.

There are likely many complexities but as this begins to stabilize it might be worth considering ways to 'lock-in' the behavior to help detect any bugs that could creep in.

Otherwise looks great! 🙂

File: Gems/MotionMatching/Code/Source/FeatureVelocity.cpp (hunk 1 above)
Comment (amzn-jillich) [reply Gems/MotionMatching/Code/Source/FeatureVelocity.cpp:74]: Great suggestion, will get this in.

**Review by amzn-rhhong — APPROVED** [review]

**Pull request merged**